lib/snippets/kubernetes-deploy

```diff
@@ -66,99 +200,455 @@ class KubernetesDeploy
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class PersistentVolumeClaim < KubernetesResource
+    TIMEOUT = 60 * 5
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      @status, st = run_kubectl("get", type, @name, "--output=jsonpath={.status.phase}")
+      @found = st.success?
+      log_status
+    end
+
+    def deploy_succeeded?
+      @status == "Bound"
+    end
+
+    def deploy_failed?
+      @status == "Lost"
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Ingress < KubernetesResource
+    TIMEOUT = 30
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @status = st.success? ? "Created" : "Unknown"
+      @found = st.success?
+      log_status
+    end
+
+    def deploy_succeeded?
+      exists?
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+
+    def group_name
+      "Ingresses"
+    end
+  end
+
+  class Service < KubernetesResource
+    TIMEOUT = 60 * 15
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @found = st.success?
+      if @found
+        endpoints, st = run_kubectl("get", "endpoints", @name, "--output=jsonpath={.subsets[*].addresses[*].ip}")
+        @num_endpoints = (st.success? ? endpoints.split.length : 0)
+      else
+        @num_endpoints = 0
+      end
+      @status = "#{@num_endpoints} endpoints"
+      log_status
+    end
+
+    def deploy_succeeded?
+      @num_endpoints > 0
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Pod < KubernetesResource
+    TIMEOUT = 60 * 15
+    SUSPICIOUS_CONTAINER_STATES = %w(ImagePullBackOff RunContainerError).freeze
+
+    def initialize(name, namespace, file, parent: nil)
+      @name, @namespace, @file, @parent = name, namespace, file, parent
+      @bare = !@parent
+    end
+
+    def sync
+      out, st = run_kubectl("get", type, @name, "-a", "--output=json")
+      if @found = st.success?
+        pod_data = JSON.parse(out)
+        interpret_json_data(pod_data)
+      else # reset
+        @status = @phase = nil
+        @ready = false
+        @containers = []
+      end
+      display_logs if @bare && deploy_finished?
+      log_status
+    end
+
+    def interpret_json_data(pod_data)
+      @phase = (pod_data["metadata"]["deletionTimestamp"] ? "Terminating" : pod_data["status"]["phase"])
+      @containers = pod_data["spec"]["containers"].map { |c| c["name"] }
+
+      if @deploy_started && pod_data["status"]["containerStatuses"]
+        pod_data["status"]["containerStatuses"].each do |status|
+          reason = status.dig("state", "waiting", "reason")
```

Comment:
It's a ruby 2.3.1 thing... actually, our Shipit is on 2.3.1, but that doesn't mean everyone's is. It seems like the engine's official requirement is 2.2.2 though (https://github.com/Shopify/shipit-engine/blob/de0aeeaa4920a520ff0ec4a3b228795adc3dd0ec/template.rb#L3), so I actually shouldn't use this!